### ISSUE TITLE
ArgumentError: unknown registry

### ISSUE THREAD

**Issue by tomtaylor** [sneako/finch#277]
We're seeing semi-regular stack traces in Sentry with the following:

```
ArgumentError: unknown registry: Obfuscated.FinchPool
  File "lib/registry.ex", line 1391, in Registry.info!/1
  File "lib/registry.ex", line 1007, in Registry.register/3
  File "lib/finch/http2/pool.ex", line 207, in Finch.HTTP2.Pool.init/1
  File "gen_statem.erl", line 984, in :gen_statem.init_it/6
  File "proc_lib.erl", line 241, in :proc_lib.init_p_do_apply/3
```

I think this is happening during our deployments when the app is being shutdown. Is it possible that the process tree in pool is being shutdown in the wrong order?

We're running finch `0.18.0` and mint `1.6.1` on Elixir 1.17.2. I think this has started happening since the upgrade to Elixir 1.17.

**Comment by sneako** [comment]
I don't think that sounds likely. My best guess would be that your application is still trying to send requests after Finch has already been stopped. Does that sound possible?

**Comment by sneako** [comment]
Closing due to inactivity, but if you have a more information that seems to contradict my theory feel free to comment.

**Comment by elfenlaid** [comment]
Same for me, the same stack trace. This happens only on startups. I've never seen it on a local machine, but AppSignal reports this crashes regularly on production servers.

I assume it happens on startups because AppSignal attaches a hash of deployed changes to reported errors. The first instance of this error happens right after starting up a new version.

Happens after upgrading to Elixir/OTP 1.17.2/26.2.5.2 from 1.16.2/26.1.1. We were using Finch 0.18.0 for some time before upgrading the language version without errors.

Just to be certain nothing uses Finch before it's initialized, I start it up as the first application child. Still was getting the error. My configuration looks like this:

```elixir
    children = [
      # Start Finch
      {Finch,
       name: Mindme.Finch,
       pools: %{
         "https://api.stripe.com" => [protocols: [:http2]]
       }},
```

It could not even be related to Finch, we see similar errors start to pop up with other dependencies. So far, it happens to Mongo as well:

```
lib/gen_server.ex:1128 GenServer.call/3	
lib/mongo/auth.ex:34 Mongo.Auth.credentials/1	
lib/mongo/auth.ex:20 Mongo.Auth.run/2	
lib/mongo/mongo_db_connection.ex:131 Mongo.MongoDBConnection.connect/2	
lib/db_connection/connection.ex:74 DBConnection.Connection.handle_event/4	
gen_statem.erl:1397 :gen_statem.loop_state_callback/11	
proc_lib.erl:251 :proc_lib.wake_up/3
```

Though, in this case it starts in `proc_lib.erl:251 :proc_lib.wake_up/3`, while for Finch it starts in `proc_lib.erl:241 :proc_lib.init_p_do_apply/3`.

**Comment by wojtekmach** [comment]
There is a similar elixir forum thread: https://elixirforum.com/t/finch-unknown-registry-crash-on-application-start-up/65609/1

@tomtaylor @elfenlaid perhaps you're configuring a dependency to use a Finch pool that your start in your app?

**Comment by elfenlaid** [comment]
@wojtekmach Oh, thanks for linking the topic!

> perhaps you're configuring a dependency to use a Finch pool that your start in your app?

Unfortunately, I don't think it's the case. Only my app is dependent on Finch

**Comment by sneako** [comment]
Fix released in [v0.19.0](https://hex.pm/packages/finch/0.19.0), thanks everyone!

**Comment by tomtaylor** [comment]
Thanks @sneako!
